adapters/commons/src/main/java/io/vanillabp/cockpit/adapter/common/workflow/kafka/WorkflowKafkaPublishing.java

```diff
@@ -0,0 +1,152 @@
+
+            var event = workflowMapper.map(workflowCancelledEvent);
+            this.sendWorkflowEvent(event.getClass(), event.toByteArray());
+
+        } else if(eventObject instanceof WorkflowCompletedEvent workflowCompletedEvent) {
+
+            var event = workflowMapper.map(workflowCompletedEvent);
+            this.sendWorkflowEvent(event.getClass(), event.toByteArray());
+
+        }
+        // else if suspended
+        // else if activated
+        else {
+            throw new RuntimeException(
+                    "Unsupported event type '"
+                            + eventObject.getClass().getName()
+                            + "'!");
+        }
+    }
+
+    private void editWorkflowCreatedOrUpdatedEvent(WorkflowCreatedEvent eventObject) {
+        final var event = eventObject;
+        event.setSource(workerId);
+
+        final var commonWorkflowsProperties = workflowsCockpitProperties
+                .getWorkflows()
+                .stream()
+                .filter(props -> props.matches(event.getWorkflowModule()))
+                .findFirst()
+                .get();
+        final var workflowsProperties = workflowsCockpitProperties
+                .getWorkflows()
+                .stream()
+                .filter(props -> !props.matches(event.getWorkflowModule()))
+                .filter(props -> props.matches(event.getWorkflowModule(), event.getBpmnProcessId()))
+                .findFirst()
+                .get();
+
+        event.setWorkflowProviderApiUriPath(
+                StringUtils.hasText(workflowsProperties.getWorkflowProviderApiPath())
+                        ? workflowsProperties.getWorkflowProviderApiPath()
+                        : commonWorkflowsProperties.getWorkflowProviderApiPath());
+        var uiUriPath = workflowsProperties.getUiUriPath();
+        if (!StringUtils.hasText(uiUriPath)) {
+            uiUriPath = commonWorkflowsProperties.getUiUriPath();
+        }
+        if (!StringUtils.hasText(uiUriPath)) {
+            uiUriPath = event.getUiUriPath();
+        } else if (StringUtils.hasText(event.getUiUriPath())) {
+            uiUriPath += event.getUiUriPath();
+        }
+        event.setUiUriPath(uiUriPath.replace("//", "/"));
+        var uiUriType = workflowsProperties.getUiUriType();
+        if (!StringUtils.hasText(uiUriType)) {
+            uiUriType = commonWorkflowsProperties.getUiUriType();
+        }
+        if (!StringUtils.hasText(uiUriType)) {
+            uiUriType = properties.getUiUriType();
+        }
+        event.setUiUriType(WorkflowUiUriType.fromValue(uiUriType));
+        event.setWorkflowModuleUri(
+                StringUtils.hasText(workflowsProperties.getWorkflowModuleUri())
+                        ? workflowsProperties.getWorkflowModuleUri()
+                        : commonWorkflowsProperties.getWorkflowModuleUri());
+    }
+
+
+    private void sendWorkflowEvent(Class<?> eventClass, byte[] eventStream) {
+        kafkaTemplate.send(
```

Comment:
The API-version used should be sent as a Kafka-header to be able to do semantic upcasting if necessary. For REST this is done implicitly by using different endpoints for different versions.